enclave/src/openfinex/openfinex_client.rs

```diff
@@ -0,0 +1,382 @@
+            received_plaintext: Vec::new(),
+            sendable_plaintext: Vec::new(),
+            response_handler,
+        }
+    }
+
+    fn jwt_handshake(&mut self) {
+        //FIXME: this should probably some proper hostname
+        let ip = if self.uri.ip() == "127.0.0.1" {
+            "localhost".to_owned()
+        } else {
+            self.uri.ip()
+        };
+        let host = format!("Host: {}:{}\r\n", ip, self.uri.port());
+        let authorization = format!(
+            "Authorization: Bearer {}\r\n",
+            jwt::create_jwt("U487205863".to_owned())
+        );
+        let ws_version =
+            "Sec-WebSocket-Version: 13\r\nSec-WebSocket-Key: s3pPLMBiTxaQ9kYGzzhZRbK+xOo=\r\n";
+        let connection = "Connection: Upgrade\r\nUpgrade: websocket\r\n";
+        let data = format!(
+            "GET {} HTTP/1.1\r\n{}{}{}{}\r\n",
+            self.uri.path(),
+            ws_version,
+            connection,
+            authorization,
+            host
+        );
+        debug!("Sending http request: {}", data);
+        if let Err(e) = self.socket.write(data.as_bytes()){
+            error!("Could not handshake with openfinex server: {}", e);
+        };
+    }
+
+    fn wants_write(&self) -> bool {
+        !self.sendable_plaintext.is_empty()
+    }
+
+    // we always want to read except when we already received some text
+    // we want to process
+    fn wants_read(&self) -> bool {
+        self.received_plaintext.is_empty()
+    }
+
+    /* fn parse_http_response(&mut self, buffer: &[u8]) -> c_int {
+        // split header and body
+        let mut index = 0;
+        let mut _prev_byte = 0;
+        let new_line = 0;
+        // Header - Body seperator: [13, 10, 13, 10] = \r\n\r\n
+        // start with header
+        for byte in buffer.iter() {
+            if *byte == 129 {
+                break;
+            } else if *byte == 0 {
+                // EOF
+                break;
+            }
+            index += 1;
+            _prev_byte = *byte;
+        }
+        let header = String::from_utf8_lossy(&buffer[0..index]);
+        debug!("Header: {}", header);
+        // get body
+        let body = String::new();
+        // TODO if necessary
+        debug!("Body: {}", body);
+        1
+    } */
+
+    /// We're ready to do a read.
+    fn do_read(&mut self) -> io::Result<usize> {
+        //FIXME: maybe we can use up buffer read here?
+        let mut buffer = [0 as u8; 512]; // Dummy buffer. will not be necessary with tls client
+        let rc = self.socket.read(&mut buffer);
+        if rc.is_err() {
+            error!("TLS read error: {:?}", rc);
+            //self.closing = true;
+            return rc;
+        }
+
+        // If we're ready but there's no data: EOF.
+        if rc.unwrap() == 0 {
+            error!("EOF");
+            //self.closing = true;
+            //self.clean_closure = true;
+            return Ok(0);
+        }
+
+        if buffer[0] == 72 {
```

Comment:
This checks for an "H" within the response (for HTTP). A websocket response will always start with 0000 or 1000, never with 72 = 1001000